File: solutions/846/main.py
from typing import List 

class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        hand.sort()
        visited = [0]*len(hand)
        for i in range(len(hand)-groupSize+1):
            cardIdx = 1
            if visited[i] == 1:
                continue
            cards = []
            cards.append(i)
            j = i + 1
            while j < len(hand) and cardIdx != groupSize:
                if visited[j] == 0 and hand[j] == hand[cards[-1]] + 1:
                    cards.append(j)
                    cardIdx += 1
                j += 1
            if cardIdx == groupSize:
                for c in cards:
                    visited[c] = 1
            else:
                return False
        if sum(visited) != len(hand):
            return False
        return True

if __name__ == "__main__": 
    hand = [1,2,3,6,2,3,4,7,8]
    groupSize = 3
    s = Solution()
    assert s.isNStraightHand(hand, groupSize) 

File: solutions/846/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Thu Dec 30 15:30:01 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <algorithm> 
#include <numeric> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        sort(hand.begin(), hand.end());
        vector<int> visited(hand.size(), 0);
        for (int i = 0; i < hand.size()-groupSize+1; ++i) {
            if (visited[i] == 1) continue;
            int cardId = 1;
            int j = i + 1;
            vector<int> cards{i};
            while (j < hand.size() && cardId != groupSize) {
                if (visited[j] != 1 && hand[j] == hand[cards[cards.size()-1]]+1) {
                    cardId++;
                    cards.push_back(j);
                }
                j++;
            }
            if (cardId != groupSize) return false;
            for (auto& c : cards) visited[c] = 1;
        }
        if (accumulate(visited.begin(), visited.end(), 0) != hand.size()) return false;
        return true;
    }
};

TEST(Test846, SimpleTest) {
    vector<int> hand{
        1,2,3,6,2,3,4,7,8
    };
    const int groupSize = 3;
    Solution s; 
    EXPECT_TRUE(s.isNStraightHand(hand, groupSize));
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
